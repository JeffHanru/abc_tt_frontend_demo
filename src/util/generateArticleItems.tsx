import {
  ArticleMedia,
  ArticleText,
  ArticleTitle,
  IframeMedia,
} from "../components/Share";
import { NormalTextProps, QuoteTextProps, Text, MediaEmbedded } from "../types";

export const generateArticleItems = (
  text: Text,
  mediaEmbedded: MediaEmbedded[]
) => {
  return (
    <>
      {text.json.children.map((element) => {
        const { tagname, children } = element;

        if (tagname === "p") {
          const componentInfo = children[0] as NormalTextProps;
          return <ArticleText>{componentInfo.content}</ArticleText>;
        }

        if (tagname === "pullquote") {
          const componentInfo = children[0] as QuoteTextProps;
          return (
            <ArticleText variant="quote">
              {componentInfo.children[0].content}
            </ArticleText>
          );
        }

        if (tagname === "a") {
          const embedMedia = mediaEmbedded.find(
            (media) => media.id === element.parameters?.ref
          );
          console.log(embedMedia);
          if (
            embedMedia?.externalembed &&
            embedMedia.viewType === "instagram"
          ) {
            return <IframeMedia src={embedMedia.externalembed.url} />;
          }
          if (embedMedia && (embedMedia.docType === "ImageProxy" || embedMedia.docType === "Image")) {
            const contentText = embedMedia.byLine?.json.children[0].children
              ? embedMedia.byLine.json.children[0].children[0].content
              : "";

            const embedMediaInfo = {
              src: embedMedia.media?.image.primary.complete[0].url,
              caption: embedMedia.caption,
              suppliedText: contentText,
              align: element.parameters?.align,
            };
            return <ArticleMedia {...embedMediaInfo} />;
          }
          return null;
        }

        if (tagname === "h1") {
          const componentInfo = children[0] as NormalTextProps;

          return <ArticleTitle>{componentInfo.content}</ArticleTitle>;
        }
        return "otherwise default";
      })}
    </>
  );
};
