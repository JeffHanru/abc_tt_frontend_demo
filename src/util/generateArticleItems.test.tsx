import { generateArticleItems } from "./generateArticleItems";
// eslint-disable-next-line jest/no-mocks-import
import {
  sampleArticleData,
  sampleImageData,
  sampleImageMedia,
  sampleInstagramEmbedMedia,
  sampleInstagrameData,
  sampleQuoteData,
  sampleTeaserData,
  sampleTeaserMedia,
  sampleNoMatchData,
  sampleNonExistTypeData
} from "./__mocks__/data";

jest.mock("../components/Share", () => {
  return {
    ArticleMedia: () => "ArticleMedia",
    ArticleText: () => "ArticleText",
    ArticleTitle: () => "ArticleTitle",
    InstagramIframeMedia: () => "InstagramIframeMedia",
    TeaserMedia: () => "TeaserMedia",
  };
});

const mediaData = [sampleImageMedia, sampleInstagramEmbedMedia, sampleTeaserMedia]

describe("generateArticleItems", () => {
  // Since all actually rendered component in this method have their specific unit test,
  // It is safe to do snapshot test
  it("renders the ArticleText content as expected", () => {
    const result = generateArticleItems(sampleArticleData, mediaData);
    expect(result).toMatchSnapshot();
  });

  it("renders the QuoteText content as expected", () => {
    const result = generateArticleItems(sampleQuoteData, mediaData);
    expect(result).toMatchSnapshot();
  });

  it("renders the image media content as expected", () => {
    const result = generateArticleItems(sampleImageData, mediaData);
    expect(result).toMatchSnapshot();
  });

  it("renders the instagram media content as expected", () => {
    const result = generateArticleItems(sampleInstagrameData, mediaData);
    expect(result).toMatchSnapshot();
  });

  it("renders the teaser media content as expected", () => {
    const result = generateArticleItems(sampleTeaserData, mediaData);
    expect(result).toMatchSnapshot();
  });

  it("renders no match media content as expected", () => {
    const result = generateArticleItems(sampleNoMatchData, mediaData);
    expect(result).toMatchSnapshot();
  });

  it("renders non exist type data as expected", () => {
    const result = generateArticleItems(sampleNonExistTypeData, mediaData);
    expect(result).toMatchSnapshot();
  });
});
