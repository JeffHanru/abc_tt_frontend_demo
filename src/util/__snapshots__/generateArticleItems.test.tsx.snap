// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateArticleItems renders no match media content as expected 1`] = `<React.Fragment />`;

exports[`generateArticleItems renders non exist type data as expected 1`] = `<React.Fragment />`;

exports[`generateArticleItems renders the ArticleText content as expected 1`] = `
<React.Fragment>
  <ArticleText>
    Women make up only 15 per cent of engineers with lack of support for professional development and gender discrimination still seen as barriers for women in engineering.
  </ArticleText>
</React.Fragment>
`;

exports[`generateArticleItems renders the QuoteText content as expected 1`] = `
<React.Fragment>
  <ArticleText
    variant="quote"
  >
    "I've even been lucky enough to work at a site that had Spanish coins on it."
  </ArticleText>
</React.Fragment>
`;

exports[`generateArticleItems renders the image media content as expected 1`] = `
<React.Fragment>
  <ArticleMedia
    align="embed"
    caption="Maddy McAllister manages the collections at Townsville's Museum of Tropical Queensland."
    src="https://live-production.wcms.abc-cdn.net.au/89d0a84c3dbef91e007ee42ed029935e?impolicy=wcms_crop_resize&cropH=3333&cropW=5000&xPos=0&yPos=0&width=862&height=575"
    suppliedText="ABC North Queensland: Zilla Gordon"
  />
</React.Fragment>
`;

exports[`generateArticleItems renders the instagram media content as expected 1`] = `
<React.Fragment>
  <InstagramIframeMedia
    src="https://www.instagram.com/p/CPDPd6jDJQK/"
  />
</React.Fragment>
`;

exports[`generateArticleItems renders the teaser media content as expected 1`] = `
<React.Fragment>
  <TeaserMedia
    text="Tell us your location and find more local ABC News and information"
    title="Find more local news"
    url="https://www.abc.net.au/local/"
  />
</React.Fragment>
`;
